src/wrappers.jl

```diff
@@ -0,0 +1,258 @@
+    @unpack inner, outer, isleft = v
+end
+
+## Krylov.jl
+
+struct KrylovJL{F,Tl,Tr,T,I,A,K} <: AbstractKrylovSubspaceMethod
+    KrylovAlg::F
+    Pl::Tl
+    Pr::Tr
+    abstol::T
+    reltol::T
+    maxiter::I
+    ifverbose::Bool
+    gmres_restart::I
+    window::I
+    args::A
+    kwargs::K
+end
+
+function KrylovJL(args...; KrylovAlg = Krylov.gmres!, Pl=I, Pr=I,
+                  abstol=0.0, reltol=0.0, maxiter=0, ifverbose=false,
+                  gmres_restart=20, window=0,              # for building solver
+                  kwargs...)
+
+    return KrylovJL(KrylovAlg, Pl, Pr, abstol, reltol, maxiter, ifverbose,
+                    gmres_restart, window,
+                    args, kwargs)
+end
+
+KrylovJL_CG(args...;kwargs...) =
+    KrylovJL(args...; KrylovAlg=Krylov.cg!, kwargs...)
+KrylovJL_GMRES(args...;kwargs...) =
+    KrylovJL(args...; KrylovAlg=Krylov.gmres!, kwargs...)
+KrylovJL_BICGSTAB(args...;kwargs...) =
+    KrylovJL(args...; KrylovAlg=Krylov.bicgstab!, kwargs...)
+KrylovJL_MINRES(args...;kwargs...) =
+    KrylovJL(args...; KrylovAlg=Krylov.minres!, kwargs...)
+
+const KrylovJL_solvers = Dict(
+  (Krylov.lsmr!       => Krylov.LsmrSolver          ),
+  (Krylov.cgs!        => Krylov.CgsSolver           ),
+  (Krylov.usymlq!     => Krylov.UsymlqSolver        ),
+  (Krylov.lnlq!       => Krylov.LnlqSolver          ),
+  (Krylov.bicgstab!   => Krylov.BicgstabSolver      ),
+  (Krylov.crls!       => Krylov.CrlsSolver          ),
+  (Krylov.lsqr!       => Krylov.LsqrSolver          ),
+  (Krylov.minres!     => Krylov.MinresSolver        ),
+  (Krylov.cgne!       => Krylov.CgneSolver          ),
+  (Krylov.dqgmres!    => Krylov.DqgmresSolver       ),
+  (Krylov.symmlq!     => Krylov.SymmlqSolver        ),
+  (Krylov.trimr!      => Krylov.TrimrSolver         ),
+  (Krylov.usymqr!     => Krylov.UsymqrSolver        ),
+  (Krylov.bilqr!      => Krylov.BilqrSolver         ),
+  (Krylov.cr!         => Krylov.CrSolver            ),
+  (Krylov.craigmr!    => Krylov.CraigmrSolver       ),
+  (Krylov.tricg!      => Krylov.TricgSolver         ),
+  (Krylov.craig!      => Krylov.CraigSolver         ),
+  (Krylov.diom!       => Krylov.DiomSolver          ),
+  (Krylov.lslq!       => Krylov.LslqSolver          ),
+  (Krylov.trilqr!     => Krylov.TrilqrSolver        ),
+  (Krylov.crmr!       => Krylov.CrmrSolver          ),
+  (Krylov.cg!         => Krylov.CgSolver            ),
+  (Krylov.cg_lanczos! => Krylov.CgLanczosShiftSolver),
+  (Krylov.cgls!       => Krylov.CglsSolver          ),
+  (Krylov.cg_lanczos! => Krylov.CgLanczosSolver     ),
+  (Krylov.bilq!       => Krylov.BilqSolver          ),
+  (Krylov.minres_qlp! => Krylov.MinresQlpSolver     ),
+  (Krylov.qmr!        => Krylov.QmrSolver           ),
+  (Krylov.gmres!      => Krylov.GmresSolver         ),
+  (Krylov.fom!        => Krylov.FomSolver           ),
+ )
+
+function init_cacheval(alg::KrylovJL, A, b, u)
+
+    KS = KrylovJL_solvers[alg.KrylovAlg]
```

Comment:
This is going to be unstable. Mostly worried about the compile time effect. It's better to make that a function with a bunch of dispatches.

Reply:
Agreed.

Reply:
changed it to a function with if statements. have a look now?

Reply:
that should be fine. I would've done dispatch, but both will evaluate at compile time.